test/stockpile_test.clj

```diff
@@ -0,0 +1,306 @@
+            (rm-r (.getAbsolutePath qdir)))))))))
+
+(deftest contending-enqueue-dequeue-performance
+  (call-with-temp-dir-path
+   (fn [tmpdir]
+     (let [qdir (.toFile (.resolve tmpdir "queue"))
+           q (stock/create qdir)
+           batch-size 2000
+           start (System/nanoTime)
+           entries (seque (int (max 100 (/ batch-size 10)))
+                          (for [i (range batch-size)]
+                            (store-str q (str i))))]
+       (doall
+        (map (fn [i entry]
+               (is (= (str i) (slurp-entry q entry)))
+               (stock/discard q entry)
+               ;; Implementation detail, but worth checking for
+               ;; current implementation
+               (is (thrown? NoSuchFileException
+                            (slurp-entry q entry))))
+             (range batch-size)
+             entries))
+       (binding [*out* *err*]
+         (printf "Enqueued and dequeued %d tiny messages in parallel at %.2f/s\n"
+                 batch-size
+                 (double (/ batch-size
+                            (/ (- (System/nanoTime) start)
+                               billion))))
+         (flush))))))
+
+(deftest simple-race
+  (call-with-temp-dir-path
+   (fn [tmpdir]
+     (let [batch-size 300
+           qdir (.toFile (.resolve tmpdir "queue"))
+           q (stock/create qdir)
+           state (atom {:entries () :victim nil})
+           finished? (atom false)
+           writer (future
+                    (dotimes [i batch-size]
+                      (swap! state update :entries conj
+                             [i (store-str q (str i) (str "meta-" i))])))
+           reader (future
+                    (while (not @finished?)
+                      (let [{:keys [victim]} (swap! state
+                                                    (fn [{[v & r] :entries}]
+                                                      {:entries r
+                                                       :victim v}))
+                            [val entry] victim]
+                        (when victim
+                          (= (str val) (slurp-entry q entry))
+                          (swap! state update :entries conj victim)))))
+           discarder (future
+                       (loop [i 0]
+                         (when (< i batch-size)
+                           (let [{:keys [victim]} (swap! state
+                                                         (fn [{[v & r] :entries}]
+                                                           {:entries r
+                                                            :victim v}))
+                                 [val entry] victim]
+                             (if entry
+                               (do
+                                 (stock/discard q entry)
+                                 (recur (inc i)))
+                               (recur i))))))]
+       @writer @discarder
+       (reset! finished? true)
+       @reader))))
```

Comment:
Well there actually was one missing (is ...) wrapper (around the =), but see next comment below...